test/e2e/autoscaling/horizontal_pod_autoscaling_behavior.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package autoscaling
+
+import (
+	"time"
+
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2eautoscaling "k8s.io/kubernetes/test/e2e/framework/autoscaling"
+
+	"github.com/onsi/ginkgo"
+)
+
+var _ = SIGDescribe("[Feature:HPA] [Serial] [Slow] Horizontal pod autoscaling (non-default behavior)", func() {
+	f := framework.NewDefaultFramework("horizontal-pod-autoscaling")
+
+	ginkgo.Describe("with low downscale stabilization window", func() {
+		ginkgo.It("should scale down soon after the stabilization period", func() {
+			ginkgo.By("setting up resource consumer and HPA")
+			podCPURequest := 500
+			targetCPUUtilizationPercent := 25
+			usageForSingleReplica := 110
+			initPods := 1
+			initCPUUsageTotal := initPods * usageForSingleReplica
+			downScaleStabilization := 1 * time.Minute
+
+			rc := e2eautoscaling.NewDynamicResourceConsumer(
+				"consumer", f.Namespace.Name, e2eautoscaling.KindDeployment, initPods,
+				initCPUUsageTotal, 0, 0, int64(podCPURequest), 200,
+				f.ClientSet, f.ScalesGetter, e2eautoscaling.Disable, e2eautoscaling.Idle,
+			)
+			defer rc.CleanUp()
+
+			hpa := e2eautoscaling.CreateCPUHorizontalPodAutoscalerWithBehavior(
+				rc, int32(targetCPUUtilizationPercent), 1, 5, int32(downScaleStabilization.Seconds()),
+			)
+			defer e2eautoscaling.DeleteHPAWithBehavior(rc, hpa.Name)
+
+			// 30s of metrics window and
+			// up to 30s until the old metrics window finishes
+			// and additional 15s until metrics available on a new pod
+			metricsAvailableDelay := 75 * time.Second
+			hpaReconciliationInterval := 15 * time.Second
+			actuationDelay := 10 * time.Second
+			maxHPAReactionTime := metricsAvailableDelay + hpaReconciliationInterval + actuationDelay
+
+			maxConsumeCPUDelay := 30 * time.Second
+			waitForReplicasPollInterval := 20 * time.Second
+			maxResourceConsumerDelay := maxConsumeCPUDelay + waitForReplicasPollInterval
+
+			waitBuffer := 1 * time.Minute
+
+			ginkgo.By("triggering scale up to record a recommendation")
+			rc.ConsumeCPU(3 * usageForSingleReplica)
+			rc.WaitForReplicas(3, maxHPAReactionTime+maxResourceConsumerDelay+waitBuffer)
```

Comment:
Why the test starts with scale up instead of stating with higher number of relpicas?

Reply:
I want to make sure HPA has a recorded recommendation for stabilization, that it doesn't scale down immediately in the first reconciliation cycle (ignoring the stabilization logic). I guess I could instead make a 1m sleep after creating the HPA and ensure the starting replicas stay unchanged after the sleep period. I thought making a scale up and then scale down was a more natural setup and serves as a sanity check that HPA is ready and doing its job. WDYT?

Reply:
Sounds good to me. I think it would be good to add this explanation to the test